From 57dc50af1561a04e7d41747e7dbacd6d72b8d866 Mon Sep 17 00:00:00 2001
From: Marcelo Tosatti <mtosatti@redhat.com>
Date: Thu, 22 May 2014 13:16:58 -0400
Subject: [virt] kvm/x86: expose invariant tsc cpuid bit

Message-id: <20140522131657.GA25212@amt.cnet>
Patchwork-id: 80261
O-Subject: [RHEL7.1 PATCH] KVM: x86: expose invariant tsc cpuid bit (v2)
Bugzilla: 1095990
RH-Acked-by: Eduardo Habkost <ehabkost@redhat.com>
RH-Acked-by: Laszlo Ersek <lersek@redhat.com>
RH-Acked-by: Andrew Jones <drjones@redhat.com>
RH-Acked-by: Paolo Bonzini <pbonzini@redhat.com>

Invariant TSC is a property of TSC, no additional
support code necessary.

Signed-off-by: Marcelo Tosatti <mtosatti@redhat.com>
Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
(cherry picked from commit e4c9a5a17567f8ea975bdcfdd1bf9d63965de6c9 of
kvm.git queue branch)
BZ: 1095990

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/kvm/cpuid.c b/arch/x86/kvm/cpuid.c
index 0a1e3b8..bfcd6b2 100644
--- a/arch/x86/kvm/cpuid.c
+++ b/arch/x86/kvm/cpuid.c
@@ -466,6 +466,13 @@ static int do_cpuid_ent(struct kvm_cpuid_entry2 *entry, u32 function,
   entry->ecx &= kvm_supported_word6_x86_features;
   cpuid_mask(&entry->ecx, 6);
   break;
+ case 0x80000007: /* Advanced power management */
+  /* invariant TSC is CPUID.80000007H:EDX[8] */
+  entry->edx &= (1 << 8);
+  /* mask against host */
+  entry->edx &= boot_cpu_data.x86_power;
+  entry->eax = entry->ebx = entry->ecx = 0;
+  break;
  case 0x80000008: {
   unsigned g_phys_as = (entry->eax >> 16) & 0xff;
   unsigned virt_as = max((entry->eax >> 8) & 0xff, 48U);
@@ -496,7 +503,6 @@ static int do_cpuid_ent(struct kvm_cpuid_entry2 *entry, u32 function,
  case 3: /* Processor serial number */
  case 5: /* MONITOR/MWAIT */
  case 6: /* Thermal management */
- case 0x80000007: /* Advanced power management */
  case 0xC0000002:
  case 0xC0000003:
  case 0xC0000004:
-- 
1.7.1